File: reference/embed.py
import json

import torch
from safetensors import safe_open

index = json.load(open("models/qwen3-4b-base/model.safetensors.index.json"))
file = index["weight_map"]["model.embed_tokens.weight"]

with safe_open(f"models/qwen3-4b-base/{file}", framework="pt") as f:
    w = f.get_tensor("model.embed_tokens.weight")
    print(w.shape)
    for id in [9707, 1879]:
        print(id, w[id][:4].to(torch.float32).tolist())

File: examples/embed.rs
use std::fs;

use llm::json;
use llm::safetensors::Shard;
use llm::tokenizer::Tokenizer;

fn main() {
    let text = fs::read_to_string("models/qwen3-4b-base/model.safetensors.index.json").unwrap();
    let index = json::parse(&text).unwrap();
    let file = index
        .get("weight_map")
        .and_then(|map| map.get("model.embed_tokens.weight"))
        .and_then(|file| file.as_str())
        .unwrap();
    println!("{file}");
    let shard = Shard::load(&format!("models/qwen3-4b-base/{file}")).unwrap();
    let embed = shard.tensor("model.embed_tokens.weight").unwrap();
    println!("{:?}", embed.shape);
    let tokenizer = Tokenizer::load("models/qwen3-4b-base/tokenizer.json").unwrap();
    for id in tokenizer.encode("Hello world") {
        let dim = embed.shape[1];
        println!("{id}: {:?}", &embed.data[id * dim..id * dim + 4]);
    }
}
